rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,177 @@
+
+## Details
+
+The work on OIDC joining is broken down into two parts:
+
+- Support in the Auth server for trusting an OP issued JWT. This work is general, and will be applicable to all providers.
+- Support in nodes for fetching their workload identity token from their environment. This work will be specific to each platform we intend to support.
+
+OIDC supports multiple types of token (`id_token`: a JWT encoding information about the identity, which can be verified using the issuer's public key and `access_token`: an opaque token that can be used with a `userinfo` endpoint on the issuer to obtain information about the identity). However, in the case of workload identities, `id_token` is the most prevelant. For this reason, our initial implementation will solely support `id_token`.
+
+### Auth server support
+
+We will re-use the existing endpoints around joining as much as possible. This means that the main entry-point for joining will be the existing `RegisterUsingToken` method.
+
+We will introduce a new token type, `oidc-jwt`, and add an additional field to the Token resource to allow the issuer URL to be specified (`issuer_url`).
+
+Registration flow:
+
+1. Client is configured by the user to use `oidc-jwt` joining with a specific provider. The client then uses the provider-specific logic to obtain a token.
+2. The client will call the `RegisterUsingToken` endpoint, specifying the name of the token, and providing the token that it has collected.
+3. The server will attempt to fetch the Token resource for the specified token.
+4. The server will check JWT header to ensure the `alg` is one we have allow-listed (RS[256, 384, 512])
+5. The server will check the `kid` of the JWT header, and obtain the relevant JWK from the cache or from the specified issuers well-known JWKS endpoint. It will then use the JWK to validate the token has been signed by the issuer.
+6. Other key claims of the JWT will be validated:
+  - Ensure the Issued At Time (iat) is in the past.
+  - Ensure the Expiry Time (exp) is in the future.
+7. The user's [configured Common Expression Language rule](#configuration) for the token will be evaluated against the claims, to ensure that the token is allowed to register with the Teleport cluster.
+8. Certificates will be generated for the client. The generated certificates will be non-renewable, as the client will proceed through the same steps to generate new certificates. This prevents exfiltratred credentials being used to repeatedly generate more credentials, maintaining access to the system.
+
+We will re-use `go-jose@v2` for validation of JWTs, since this library is already in use within Teleport.
+
+#### Caching JWKs
+
+Special attention should be given to the logic around caching the JWKs.
```

Comment:
will we cache in the backend or in memory?

Reply:
My gut feeling is this would be appropriate for caching in-memory, since it's a relatively small data set and it's quite cheap for us to fetch it directly from the issuer if the service restarts. I'm not too aware of the finer details of our two storage options, so if someone has a strong argument for using the backend, I'd be willing to hear it.

Reply:
One tradeoff worth at least mentioning is that with redundant auth servers, they would each have their own in-memory cache, vs being able to share it if it was in the backend. I would still say that an in-memory cache would probably make more sense for this though

Reply:
Thanks. I'll add some additional notes to specify that this will be an in-memory implementation :)

Reply:
[f731797](https://github.com/gravitational/teleport/pull/14414/commits/f731797f0164cb7433bdd921db332d6f61f6ff64)